### PR TITLE
Replace `EventContext` fields `prev_group` and `delta_ids` with field `state_group_deltas`

### PR DESCRIPTION
**Pull request by H-Shay**
Currently the class `EventContext` uses the fields `prev_group` and `delta_ids` to hold information about the state group previous to the event (or the state group previous to the group before the event) and the state delta between the previous state group and the new state group. This information was used to try and optimize storing the state at the event, as if the previous state group was known one could just store the state delta between the old state and the new.

However, this way of tracking the information could be very confusing. This PR attempts to make it less confusing, and replaces the fields `prev_group` and `delta_ids` with the field `state_group_deltas` in the hope that this formulation is a little clearer and easier to work with.

### CODE REVIEW

File: changelog.d/15233.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Replace EventContext fields prev_group and delta_ids with field state_group_deltas.
```
Comment (clokep) [inline changelog.d/15233.misc:1]:
```suggestion
Replace `EventContext` fields `prev_group` and `delta_ids` with field `state_group_deltas`.
```

File: synapse/events/snapshot.py [hunk 2]
```diff
@@ -131,8 +119,9 @@ class EventContext(UnpersistedEventContextBase):
     _state_group: Optional[int] = None
     state_group_before_event: Optional[int] = None
     _state_delta_due_to_event: Optional[StateMap[str]] = None
-    prev_group: Optional[int] = None
-    delta_ids: Optional[StateMap[str]] = None
+    state_group_deltas: Optional[
+        Dict[Tuple[Optional[int], Optional[int]], Optional[StateMap[str]]]
```
Comment (clokep) [inline synapse/events/snapshot.py:123]: I'm a bit surprised these are all optional, those cases don't seem covered in the docstring!

File: synapse/events/snapshot.py [hunk 3]
```diff
@@ -209,11 +196,10 @@ def deserialize(storage: "StorageControllers", input: JsonDict) -> "EventContext
             storage=storage,
             state_group=input["state_group"],
             state_group_before_event=input["state_group_before_event"],
-            prev_group=input["prev_group"],
+            state_group_deltas=_decode_state_group_delta(input["state_group_deltas"]),
```
Comment (clokep) [inline synapse/events/snapshot.py:199]: I'm not really sure when we serialize / deserialize these, do we need to do anything special during an update / rollout to account for if we serialize with the old `prev_group` and `delta_ids` fields, but then try to deserialize with the new code? Is that possible?

File: synapse/events/snapshot.py [hunk 4]
```diff
@@ -482,6 +477,38 @@ def _encode_state_dict(
...
+        return None
+
+    state_group_deltas = {}
+    for element in input:
+        state_group_deltas[(element[0], element[1])] = _decode_state_dict(element[2])
+
+    return state_group_deltas
+
+
 def _decode_state_dict(
```
Comment (clokep) [inline synapse/events/snapshot.py:512]: Do we still use the `_{encode,decode}_state_dict` functions?

File: synapse/events/snapshot.py [hunk 5]
```diff
@@ -106,21 +106,9 @@ class EventContext(UnpersistedEventContextBase):
...
-            will always also be ``None``.
-
-            Note that this *not* (necessarily) the state group associated with
-            ``_prev_state_ids``.
-
-        delta_ids: If ``prev_group`` is not None, the state delta between ``prev_group``
-            and ``state_group``.
+        state_group_deltas: a dict containing information about the state groups associated with
+            the event and the state delta between these groups, ie a map from
+            (prev state group, new state group) -> delta state dict
```
Comment (clokep) [inline synapse/events/snapshot.py:111]: I'm not really sure what this dictionary contains, it seems to only ever get set with two entries in it, but I don't really get how those differ. Can an event have multiple previous state groups and multiple new ones?

Comment (H-Shay) [reply synapse/events/snapshot.py:111]: It's a little confusing because an event has two state groups: the one before the event and the one after the event. Each of those state groups will also have a previous state group. If the event is a state event, then the state group after the event will be different than the state group before the event - and the state group before the event will become the prev group for the state group after the event. 
If the event is not a state event, then the state group before the event and the state group after the event are the same, and the prev group for the state group before the event becomes the prev group for the state group after the event. 
We only ever care about the prev group for the state group after the event, and the hope is that by adding this field it is clear which prev group is associated with which state group and it's state delta (as they are contained in the dict together). Let know if this answers your question - I might have gone on a tangent here.

File: synapse/events/snapshot.py (hunk 3 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:199]: If my reading of the code is correct, we serialize/de-serialize when sending events/contexts over replication, so I don't?? think there would be a situation where one replication endpoint would receive an incompatible context - contexts are ephemeral and not stored in the DB so I don't think upgrading would cause an issue here?

The only place I could see this being an issue is if we send event contexts out over federation and a Synapse running this version sends some event contexts to a Synapse running an older version (or vice-versa) but AFAICT we don't send event contexts out over federation, although I should probably ask to double-check. Please let me know if my logic isn't sound here, I am thinking through it real-time.

File: synapse/events/snapshot.py (hunk 4 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:512]: Yes they are still used in a: the en/decoding of `state_delta_due_to_event` and b: as a helper method in `_{encode, decode}_state_group_deltas`

File: synapse/events/snapshot.py (hunk 5 above)
Comment (clokep) [reply synapse/events/snapshot.py:111]: I'm sorry, but I'm not following this. I'll try to reword and see if I'm anywhere close.

An event has a state group before and after the event. Each state group also has a previous state group, so we have two pairs of state groups:

* (State group before event, prev state group before event)
* (State group after event, prev state group after event)

For a state event the state group before the event becomes the new prev state group after event:

* (N, M)
* (P, N)

For a non-state event there is no change in the state groups:

* (N, M)
* (N, M)

Hopefully that's correct?

----

> We only ever care about the prev group for the state group after the event, and the hope is that by adding this field it is clear which prev group is associated with which state group and it's state delta

Then why even store the information from before the event? It seems we previously only store the information we need, which looks a lot clearer to me. I'm guessing this is a stepping stone to something else, but I'm finding the new code more confusing, what will this empower us to do in the future?

File: synapse/events/snapshot.py (hunk 3 above)
Comment (clokep) [reply synapse/events/snapshot.py:199]: I seem to only see things like [this](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#changes-to-the-account-data-replication-streams) that say replication will pause until things are upgraded. Not sure if we've ever done a 'you must restart the world' one before (I thought we did, but I can't find it...)

File: synapse/events/snapshot.py (hunk 5 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:111]: Your understanding is correct, thanks for writing that out clearly. 

As for the second part, I think what I meant was not "we only care about the prev group for the state group after the event" what I really meant was "we only care about the prev group for the state group after the event and the state dict representing the state delta between state group before the event and the state group after the event", which is what is captured in the new field. 

Adding the `state_group_deltas` field was intended to relieve any future developers from having to parse out what the `prev_group` means, as this was also a complication when working with `UnpersistedEventContext`. I guess fundamentally the idea was to make the role of `prev_group` less confusing during the transition from `UnpersistedEventContext` --> `EventContext`. The `UnpersistedEventContext` does not yet have a state group after the event, so it only has the field `prev_group_for_state_group_before_event`. Rather than adding another field to the `UnpersistedEventContext` for `prev_group_for_state_group_after_the_event` the decision was made to add this all to the `state_group_deltas` instead, which represents the actual information we are after anyway (rather than building it from `prev_group` and `delta_ids` when persisting the event, see [here](https://github.com/matrix-org/synapse/blob/afb216c202feb143ce70c74c16fa50ca93da6157/synapse/storage/controllers/persist_events.py#L843)) and skip having to track all of this in the form of `prev_group`. 

I am sorry if this is hopelessly confusing, talking about `prev_group` was endlessly confusing for Erik and I and we thought this might be a way of saving others pain down the line. I am open if that's not the case, however!

File: synapse/events/snapshot.py (hunk 3 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:199]: I think using a similar warning like what you linked would work, as the issue is the same - the upgrade is backwards/forwards incompatible and the workers/master must all be on the same compatible version (and actually I think? this also implies a restart as don't you need to restart to upgrade?)

File: synapse/events/snapshot.py (hunk 5 above)
Comment (clokep) [reply synapse/events/snapshot.py:111]: I'm just not finding this any clearer then the code is today -- it seems to be duplicating information we already have in the `_state_group`, `state_group_before_event`, and `state_delta_due_to_event` fields?

----

Looking at the changes again here, *most* of the changes are internal to `EventContext`, but they're then also in `persist_events`, in here we:

* Have a dictionary which always has only two items in it;
* Only ever pull a single item out in the persistence code path.

This has code smells to it in my mind. If we know we only need one of those why do we store both of them? If we know there's always two, why use a dictionary?

Comment (H-Shay) [reply synapse/events/snapshot.py:111]: Totally fair - let me go back to the drawing board with this and see if I can't come up with anything clearer.

**Comment by clokep** [comment]
@H-Shay Is this pending my review? I *think* you had said in a meeting recently that you were going to look at this again, but want to ensure that I'm not blocking you!

**Comment by clokep** [comment]
We talked about this in person and @H-Shay is going to look into whether some of this code is even truly used in "real" situations or if the optimizations aren't needed. Once we figure that out we'll decide what to do with this PR.

Removing my review request for now.

**Comment by H-Shay** [comment]
Right so I think this is ready for another look. 

For completeness, Erik and I logged how often the optimization that prev_group/state_group_deltas enables was hit and determined it was worth keeping. Thus I have attempted here to rework the PR to make the code clearer by reducing the amount of `None`s flying around, better documentation, more documentation closer to the relevant code, and some encapsulation. Let me know what you think.

File: synapse/events/snapshot.py [hunk 6]
```diff
@@ -106,21 +106,15 @@ class EventContext(UnpersistedEventContextBase):
...
-            and ``state_group``.
+        state_group_deltas: if it exists, this is a dict collecting a mapping of two possible
+            sets of state groups and their deltas: the state group before the event and
+            after the event and the state map that represents the state group between the
+            two, i.e. (state_group_2, state_group_1) -> (state_key, event_type) -> event_id
+            or/and the state group preceding the state group before the event and the state
+            group before the event and the state map that represents the delta between those
+            two state groups.
+            This information is collected and stored as part of an optimization for persisting
+            events.
```
Comment (clokep) [inline synapse/events/snapshot.py:117]: I'm still finding this really hard to understand, I think it needs some whitespace and less complex sentences. I've rewritten it, but I don't know if it is accurate.

```suggestion
        state_group_deltas: if it exists, this is a dict collecting a mapping of the state
            difference between state groups.

            The keys are a tuple of two integers: the initial group and final state group.
            The corresponding value is a state map representing the state delta between
            these state groups.

            The dictionary is expected to have at most two entries with state groups of:

            1. The state group before the event and after the event.
            2. The state group preceding the state group before the event and the
               state group before the event.

            This information is collected and stored as part of an optimization for persisting
            events.
```

File: synapse/events/snapshot.py [hunk 7]
```diff
@@ -447,26 +425,27 @@ async def persist(self, event: EventBase) -> EventContext:
...
         # otherwise if it is a state event we need to get a state group for it
         else:
-            state_group_after_event = await self._storage.state.store_state_group(
+            self.state_group_after_event = await self._storage.state.store_state_group(
                 event.event_id,
                 event.room_id,
                 prev_group=self.state_group_before_event,
                 delta_ids=self.state_delta_due_to_event,
                 current_state_ids=None,
             )
```
Comment (clokep) [inline synapse/events/snapshot.py:438]: Kind of seems like we always should have been setting these?

File: synapse/events/snapshot.py [hunk 8]
```diff
@@ -490,3 +497,47 @@ def _decode_state_dict(
         return None
 
     return immutabledict({(etype, state_key): v for etype, state_key, v in input})
+
+
+def _build_state_group_deltas(
+    unpersisted_ctx: UnpersistedEventContext,
+) -> Dict[Tuple[int, int], StateMap]:
```
Comment (clokep) [inline synapse/events/snapshot.py:504]: This feels like it should be a method on `UnpersistedEventContext`?

File: synapse/events/snapshot.py [hunk 9]
```diff
@@ -482,6 +461,34 @@ def _encode_state_dict(
...
+    if not input:
+        return None
+
+    state_group_deltas = {}
+    for element in input:
+        state_map = _decode_state_dict(element[2])
+        assert state_map is not None
+        state_group_deltas[(element[0], element[1])] = state_map
+
+    return state_group_deltas
```
Comment (clokep) [inline synapse/events/snapshot.py:489]: Please move these above `_encode_state_dict` (so that the pairs of encode/decode functions are together).

File: synapse/events/snapshot.py [hunk 10]
```diff
@@ -482,6 +461,34 @@ def _encode_state_dict(
...
+    input: Optional[List[Tuple[int, int, List[Tuple[str, str, str]]]]]
+) -> Optional[Dict[Tuple[int, int], StateMap[str]]]:
+    if not input:
+        return None
+
+    state_group_deltas = {}
+    for element in input:
+        state_map = _decode_state_dict(element[2])
+        assert state_map is not None
+        state_group_deltas[(element[0], element[1])] = state_map
```
Comment (clokep) [inline synapse/events/snapshot.py:487]:
```suggestion
    for state_group_1, state_group_2, state_dict in input:
        state_map = _decode_state_dict(state_dict)
        assert state_map is not None
        state_group_deltas[(state_group_1, state_group_2)] = state_map
```

File: synapse/events/snapshot.py [hunk 11]
```diff
@@ -131,8 +125,7 @@ class EventContext(UnpersistedEventContextBase):
     _state_group: Optional[int] = None
     state_group_before_event: Optional[int] = None
     _state_delta_due_to_event: Optional[StateMap[str]] = None
-    prev_group: Optional[int] = None
-    delta_ids: Optional[StateMap[str]] = None
+    state_group_deltas: Optional[Dict[Tuple[int, int], StateMap[str]]] = None
```
Comment (clokep) [inline synapse/events/snapshot.py:128]: I am left wondering if this should also be a `Dict` that's just empty sometimes? 🤷

**Review by clokep — COMMENTED** [review]
I think this is massively clearer without all the `Nones` floating around! I think most of my comments are minor refactoring.

File: synapse/events/snapshot.py (hunk 3 above)
Comment (clokep) [reply synapse/events/snapshot.py:201]: I guess it would be a huge pain to have a backwards compat path since we would also need a way to *roll back* when doing this, which would mean landing the ability to decode the newer map first, waiting a version or two and then landing the new code.

tl;dr Can we document what needs to be done in the upgrade notes?

File: tests/test_state.py [hunk 12]
```diff
@@ -556,9 +556,16 @@ def test_annotate_with_old_state(
         )
 
         self.assertIsNotNone(context.state_group_before_event)
```
[the remark this answers is not in the document]
Comment (H-Shay) [reply tests/test_state.py:558]: I didn't even catch that it was repeated! Your suggestion is definitely less confusing.

File: synapse/events/snapshot.py (hunk 7 above)
Comment (clokep) [reply synapse/events/snapshot.py:444]: (There's no action to take here, just weird that we didn't previously set these on `self`.)

Comment (H-Shay) [reply synapse/events/snapshot.py:444]: We were always tracking the value, just not storing it in self - I think I changed this just to have all the info in `self` so it could be passed to `_build_state_group_deltas` - either way the value is basically immediately used to create the `EventContext` on line 442 and is then never accessed again. But I do agree this way is more consistent and thus easier to follow.

File: synapse/events/snapshot.py (hunk 8 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:510]: So I put this outside `UnpersistedEventContext` because two methods use it: `batch_persist_unpersisted_contexts`, which is a method on the class `UnpersistedEventContext`, and `persist` which is an instance method on the same class, and I couldn't find a way to easily allow both methods to use `_build_state_group_deltas` other than defining it as a class-level method, which seemed more awkward than having it defined as a helper function outside the class. Let me know which sounds better to you!

File: synapse/events/snapshot.py (hunk 6 above)
Comment (H-Shay) [reply synapse/events/snapshot.py]: Much clearer, thank you!

File: synapse/events/snapshot.py (hunk 11 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:134]: I also wondered about that - in the interest of reducing `None`s I can go ahead an do that.

File: synapse/events/snapshot.py (hunk 8 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:510]: :woman_facepalming: :woman_facepalming: :woman_facepalming: 
Sorry to have missed something so obvious - I think I got tunnel vision - thank you for clarifying this.

File: synapse/events/snapshot.py (hunk 11 above)
Comment (clokep) [reply synapse/events/snapshot.py:134]: I don't think it is a big deal either way, was just a thought 🤷

Comment (clokep) [reply synapse/events/snapshot.py:134]: (It likely means you can skip some checks that it is none and just attempt to pull things out of it though?)

File: synapse/events/snapshot.py (hunk 3 above)
Comment (H-Shay) [reply synapse/events/snapshot.py:207]: I've added some words on this, let me know if it's clear enough - in particular I wonder if I was too specific about the changes.

**Review by clokep — COMMENTED** [review]
It seems I never hit submit on this review. 😢  I think you've already updated a few of these things...

File: docs/upgrade.md [hunk 13]
```diff
@@ -88,6 +88,19 @@ process, for example:
...
+## Changes to event replication 
+
+Synapse has changed the format of `EventContext` which impacts how events to be persisted
+are sent over replication. This is a forwards- and backwards-incompatible change: v1.84.0 and below workers will not be
+able to send events to v1.85.0 workers or main process over replication, and vice versa. 
+If the master is updated to v1.85.0 then the workers must be updated to v1.84.0 for event replication to 
+function.
+This will also require a restart of the upgraded workers.
+
+Once all workers are upgraded to v1.85.0 and restarted event replication will be able to resume.
```
[the remark this answers is not in the document]
Comment (H-Shay) [reply docs/upgrade.md:102]: I actually don't know the answer to this question - thinking it through, if there were events that were unpersisted when the workers and main are upgraded and restarted, then those events would be lost due to the restart as they are not in persisted memory. Is this normally handled by the fact that workers and main are not restarted together?

[the remark this answers is not in the document]
Comment (erikjohnston) [reply docs/upgrade.md:102]: There isn't really a way at the moment to tell servers to resend.

I think the options are:

1. State in the release notes that the server must be restarted all at once on upgrade.
2. Change the serialisation format of `EventContext` to be backwards compatible. I think this means keeping the existing keys, and then adding an additional `state_deltas` key? Not sure how much work that would be.
3. Add a `v2` send events endpoint, and then:
    1. The client side tries the new endpoint, and if that fails keep retrying (until the remote worker is upgraded).
    2. Have the server accept both the current and new API (not sure how much work that is).

[the remark this answers is not in the document]
Comment (H-Shay) [reply docs/upgrade.md]: Right well given the option I'd go with 1 - it seems much easier and if that's okay then let's go with that. I've updated the upgrade notes to reflect this - let me know what you think?

**Comment by erikjohnston** [comment]
I think this is back on @H-Shay's plate?

**Comment by H-Shay** [comment]
Sooo I think I've sorted the backwards/forwards compatibility issue: serializing produces a dict which de-serializing then converts back into an object. 
The modes of possible incompatibility then are: 
1. The event context is serialized by old code and de-serialized by new code: to address that the new de-serialization code checks for the existence of the key "state_group_deltas" in the dict and if it doesn't exist just uses an empty dict for that field in the object it is building and ignores the presence of the old keys in the dict
2. The event context is serialized by the new code and de-serialized by the old code: to address this I had the new code add two "dummy" keys set to `None` corresponding to the old fields we removed ("prev_group" and "delta_ids") so the old de-serialize code will use those to reconstruct the object and ignore the new key "state_group_deltas" in the dict. 

I think this solves the problem? Let me know if my logic seems wacky.

**Review by erikjohnston — APPROVED** [review]
Woo! :tada: 

We should add a future maintenance task to remove the backwards compat hack in a couple of releases.

**Comment by H-Shay** [comment]
Merged #15233 into develop, thanks for the reviews. 

https://github.com/matrix-org/synapse/issues/15777 is the tracking issue for removing the backwards compat code in the future.
